Update Google Cloud DNS provider Rrset.Get(name) method to return a list and change the `Rrset.List()` implementation to perform a paged walk

madhusudancs opened the pull request:
Some federated service e2e tests and a few ingress tests would become flaky after a few hundred runs. @csbell spent quite a lot of time debugging this and found out that this flakiness was due to a bug in the federated service controller deletion logic. Deletion of a federated service object triggers a logic in the controller to update the DNS records corresponding to that object. This DNS record update logic would return an error in failed runs which would in-turn cause the controller to reschedule the operation. This led to an infinite retry-failure cycle that never gave the API server a chance to garbage collect the deleted service object.

A couple of days ago we started seeing a correlation between the number of resource records in a DNS managed zone and these test failures. If you look at the test runs before and after run 2900 in the test grid - https://k8s-testgrid.appspot.com/cluster-federation#gce, you will notice that the grid became super green at 2900. That's when I deleted all the dangling DNS records from the past runs.

After some investigation yesterday, we found that `ResourceRecordSet.Get()` interface and its implementation, and `ResourceRecordSet.List()` implementation at least for Google Cloud DNS were incorrect.

This PR makes minimal set of changes (read: least invasive) in Google Cloud DNS provider implementation to fix these problems:

1. Modifies DNS provider Rrset.Get(name) interface to return multiple records and updates federated service controller.

    There can be multiple DNS resource records for a given name. They can vary by type, ttl, rrdata and a number of various other parameters. It is incorrect to return a single resource record for a given name.

    This change updates the Get interface to return multiple records for a given name and uses this list in the federated service controller to perform DNS operations.

2. Update Google Cloud DNS List implementation to perform a paged walk of lists to aggregate all the DNS records.

    The current `List()` implementation just lists the DNS resorce records in a given managed zone once and retruns the list. It neither performs a paged walk nor does it consider the `page_token` in the returned response.

    This change walks all the pages and aggregates the records in the pages and returns the aggregated list. This is potentially dangerous as it can blow up memory if there are a huge number of records in the given managed zone. But this is the best we can do without changing the provider interface too much. 

    Next step is to define a new paged list interface and implement it.

**Release note**:
```release-note
NONE
```

/assign @csbell 

cc @justinsb @shashidharatd @quinton-hoole @kubernetes/sig-federation-pr-reviews

madhusudancs committed:
Modify the DNS provider Rrset.Get(name) interface to return multiple records and update federated service controller.

There can be multiple DNS resource records for a given name. They can
vary by type, ttl, rrdata and a number of various other parameters. It
is incorrect to return a single resource record for a given name.

This change updates the Get interface to return multiple records for a given
name and uses this list in the federated service controller to perform
DNS operations.

madhusudancs committed:
Update Google Cloud DNS List implementation to perform a paged walk of lists to aggregate all the DNS records.

The current `List()` implementation just lists the DNS resorce records in
a given managed zone once and retruns the list. It neither performs a paged
walk nor does it consider the `page_token` in the returned response.

This change walks all the pages and aggregates the records in the pages
and returns the aggregated list. This is potentially dangerous as it can
blow up memory if there are a huge number of records in the given
managed zone. But this is the best we can do without changing the
provider interface too much. Next step is to define a new paged list
interface and implement it.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44626)
<!-- Reviewable:end -->

madhusudancs commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

csbell commented on federation/pkg/dnsprovider/providers/google/clouddns/internal/stubs/rrsets_service.go:
```diff
@@ -31,28 +31,52 @@ type ResourceRecordSetsService struct {
+		return nil, fmt.Errorf("Zone %s not found in project %s", managedZone, project)
 	}
-	zone := s.Service.ManagedZones_.Impl[project][managedZone].(*ManagedZone)
+	return s.Service.ManagedZones_.Impl[project][managedZone].(*ManagedZone), nil
```
isn't this return z.(*ManagedZone), nil ?

madhusudancs replied:
Done.

csbell commented on federation/pkg/dnsprovider/providers/google/clouddns/rrsets.go:
```diff
@@ -30,31 +32,46 @@ type ResourceRecordSets struct {
+// List returns a list of resource records in the given project and
+// managed zone.
+// !!CAUTION!! Your memory might explode if you have a huge number of
+// records in your managed zone.
```
What is this comment based upon? We've seen a lot of recordsets in the jenkins test projects that accumulate orphaned rrsets over time but even with multiple thousands, it shouldn't be enough to OOM the controller.

madhusudancs replied:
If you are uncomfortable with this comment, I am fine removing it.

shashidharatd commented on federation/pkg/dnsprovider/providers/coredns/rrsets.go:
```diff
@@ -78,13 +77,10 @@ func (rrsets ResourceRecordSets) Get(name string) (dnsprovider.ResourceRecordSet
 		}
```
CoreDNS does support TXT record.
Since we do not write/read any other records other than A and CNAME record, it should be fine for federation. We need to make it more generic further as DNS provider will be used by many others apart from federation.

csbell replied:
This may not always be true. Either way, the code should be fixed to handle the rrset type correctly.

miekg replied:
In the [original service definition](https://github.com/coredns/coredns/blob/master/middleware/etcd/msg/service.go) there is only 1 Host field that can either be a name or an address (A or AAAA - please don't exclude AAAA from this code). So the only way to know what you got from etcd is to parse this field. TXT/MX have extra fields. 

Btw: this is adding support for etcd3? If so I would like to have support for that in upstream as well....

shashidharatd replied:
@miekg, we are still using etcd2 client for this library (same as CoreDNS). we would plan and upgrade to use etcd3 client ASAP. I think CoreDNS should be first upgraded to use etcd3 client then followed by this library.
@csbell, i would add detecting other types from etcd records in a follow-up pr.

madhusudancs replied:
@miekg thanks for confirming that. Does it make sense to add a helper to `middleware/etc/msg` package to determine the record type instead of having to re-implement it in all the clients?

@shashidharatd thanks! Please let me know when you add it.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [kubernetes/kubernetes] Update ..." ]
>shashidharatd commented on this pull request.
>
>> @@ -78,13 +77,10 @@ func (rrsets ResourceRecordSets) Get(name string) (dnsprovider.ResourceRecordSet
>
>@miekg, we are still using etcd2 client for this library (same as CoreDNS). we would plan and upgrade to use etcd3 client ASAP. I think CoreDNS should be first upgraded to use etcd3 client then followed by this library.

https://github.com/coredns/coredns/issues/341
(but not much work done)

/Miek

--
Miek Gieben

miekg commented:
[ Quoting <notifications@github.com> in "Re: [kubernetes/kubernetes] Update ..." ]
>madhusudancs commented on this pull request.
>
>
>
>> @@ -78,13 +77,10 @@ func (rrsets ResourceRecordSets) Get(name string) (dnsprovider.ResourceRecordSet
> 		}
>
>@miekg thanks for confirming that. Does it make sense to add a helper to `middleware/etc/msg` package to determine the record type instead of having to re-implement it in all the clients?

Yeah, that would make sense and cleans up my code as well.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [kubernetes/kubernetes] Update ..." ]
>madhusudancs commented on this pull request.
>
>
>
>> @@ -78,13 +77,10 @@ func (rrsets ResourceRecordSets) Get(name string) (dnsprovider.ResourceRecordSet
> 		}
>
>@miekg thanks for confirming that. Does it make sense to add a helper to `middleware/etc/msg` package to determine the record type instead of having to re-implement it in all the clients?

https://github.com/coredns/coredns/pull/627

Will be merged soonish


/Miek

--
Miek Gieben

shashidharatd commented:
Couple of issues where we might be leaking DNS records.
Looks good from CoreDNS standpoint.

---

Reviewed 5 of 12 files at r1.
Review status: 5 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 250 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-Ki9xQOzybryHvsxNrra:-Ki9xQOzybryHvsxNrrb:b-j2f8d9) ([raw file](https://github.com/kubernetes/kubernetes/blob/9c868e2f987d9d55ed4a19cc81c4cec35dc83ded/federation/pkg/federation-controller/service/dns.go#L250)):*
> ```Go
> 				glog.V(4).Infof("Existing recordset %v not equivalent to needed recordset %v removing existing and adding needed.", rrsetList, newRrset)
> 				changeSet := rrsets.StartChangeset()
> 				changeSet.Remove(found)
> ```

`found` is nil in this branch, we should be removing all the rrsets from rrsetList. otherwise we will be leaking the dns records.

---

*[federation/pkg/federation-controller/service/dns.go, line 284 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-Ki9z7aSM7i7oo1KsD8P:-Ki9z7aTCMNkd1dxycoW:bkkzdv4) ([raw file](https://github.com/kubernetes/kubernetes/blob/9c868e2f987d9d55ed4a19cc81c4cec35dc83ded/federation/pkg/federation-controller/service/dns.go#L284)):*
> ```Go
> 				// Need to replace the existing one with a better one
> 				glog.V(4).Infof("Existing recordset %v is not equivalent to needed recordset %v, removing existing and adding needed.", found, newRrset)
> 				if err = rrsets.StartChangeset().Remove(found).Add(newRrset).Apply(); err != nil {
> ```

Here too, the same comment as above.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-:-Ki9zJjmpBn2LT371Gvt:bevmub0)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
> coredns/coredns#627

> Will be merged soonish

@miekg awesome! Thank you very much!

madhusudancs commented:
Review status: 5 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 250 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-Ki9xQOzybryHvsxNrra:-KiOwUkTVL24jrR0mFYL:b-79ccux) ([raw file](https://github.com/kubernetes/kubernetes/blob/9c868e2f987d9d55ed4a19cc81c4cec35dc83ded/federation/pkg/federation-controller/service/dns.go#L250)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

`found` is nil in this branch, we should be removing all the rrsets from rrsetList. otherwise we will be leaking the dns records.
</blockquote></details>

I thought about this while writing this logic this way. But we might not be the only people adding the records for a given name. A user might have configured their own records for the same names: a record of different type or some other parameter.

---

*[federation/pkg/federation-controller/service/dns.go, line 284 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-Ki9z7aSM7i7oo1KsD8P:-KiOwguG7Lxx5Fy7EJXn:b2nreh6) ([raw file](https://github.com/kubernetes/kubernetes/blob/9c868e2f987d9d55ed4a19cc81c4cec35dc83ded/federation/pkg/federation-controller/service/dns.go#L284)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Here too, the same comment as above.
</blockquote></details>

Same as above.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
Review status: 5 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 250 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-Ki9xQOzybryHvsxNrra:-KiVCkI0epYv7XoMjAZ6:bz7oqod) ([raw file](https://github.com/kubernetes/kubernetes/blob/9c868e2f987d9d55ed4a19cc81c4cec35dc83ded/federation/pkg/federation-controller/service/dns.go#L250)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I thought about this while writing this logic this way. But we might not be the only people adding the records for a given name. A user might have configured their own records for the same names: a record of different type or some other parameter.
</blockquote></details>

@madhusudancs, it somehow feels odd that we are leaving behind stale records in some scenarios. @quinton_hoole, request your inputs here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626)*
<!-- Sent from Reviewable.io -->

madhusudancs commented on federation/pkg/dnsprovider/dns.go:
```diff
@@ -52,8 +52,12 @@ type Zone interface {
+	// Get returns the ResourceRecordSet list with the name in the Zone.
+	// This is a list because there might be multiple records of different
+	// types for a given name. If the named resource record sets do not
+	// exist, but no error occurred, the returned set will be an empty set
```
Done.

madhusudancs commented on federation/pkg/dnsprovider/providers/aws/route53/rrsets.go:
```diff
@@ -48,19 +48,30 @@ func (rrsets ResourceRecordSets) List() ([]dnsprovider.ResourceRecordSet, error)
+	err := rrsets.zone.zones.interface_.service.ListResourceRecordSetsPages(&input, func(page *route53.ListResourceRecordSetsOutput, lastPage bool) bool {
+		for _, rrset := range page.ResourceRecordSets {
+			if aws.StringValue(rrset.Name) != name {
+				return false
```
We discussed about this offline, but also mentioning here for posterity.

Route53 API takes the starting name and returns a list of records sorted on name starting from that name. Since we need exact matches on the given name and the returned list is sorted, we can terminate the walk when we reach a name that doesn't match. 

It is described here - https://docs.aws.amazon.com/sdk-for-go/api/service/route53/#Route53.ListResourceRecordSets

madhusudancs committed:
Address review comments.

madhusudancs commented:
Review status: 4 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 250 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-Ki9xQOzybryHvsxNrra:-KiVqW4pw9SDIkwd0rXm:b-8eitma) ([raw file](https://github.com/kubernetes/kubernetes/blob/9c868e2f987d9d55ed4a19cc81c4cec35dc83ded/federation/pkg/federation-controller/service/dns.go#L250)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

@madhusudancs, it somehow feels odd that we are leaving behind stale records in some scenarios. @quinton_hoole, request your inputs here.
</blockquote></details>

It is odd, not in "some scenarios", pretty much everywhere there are records not created by this controller. But let's say you have "nginx.default.myfederation.svc.example.com.", the top-level one, for which there are A records created by the controller. And the users have their own TXT records or SRV records for that, it is bad to delete them.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
Review status: 4 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 250 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-Ki9xQOzybryHvsxNrra:-KiVxC-Fw2_DEJWbJ_qE:b-s03glc) ([raw file](https://github.com/kubernetes/kubernetes/blob/9c868e2f987d9d55ed4a19cc81c4cec35dc83ded/federation/pkg/federation-controller/service/dns.go#L250)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

It is odd, not in "some scenarios", pretty much everywhere there are records not created by this controller. But let's say you have "nginx.default.myfederation.svc.example.com.", the top-level one, for which there are A records created by the controller. And the users have their own TXT records or SRV records for that, it is bad to delete them.
</blockquote></details>

@madhusudancs, i am concerned about one scenario, where the client doing discovery gets a stale 'A' Record. We are leaving the burden to client to figure out its not healthy and move on to another record.
The other scenarios, where we should not be deleting the records not created by this controller makes sense and that's probably right.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Review status: 4 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 250 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-Ki9xQOzybryHvsxNrra:-KiaRoPyYRKwXLvdpTp0:bwdbk78) ([raw file](https://github.com/kubernetes/kubernetes/blob/9c868e2f987d9d55ed4a19cc81c4cec35dc83ded/federation/pkg/federation-controller/service/dns.go#L250)):*
> where the client doing discovery gets a stale 'A' Record. We are leaving the burden to client to figure out its not healthy and move on to another record.

I am not entirely sure I understand. Could you please elaborate?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
Review status: 4 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 250 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-Ki9xQOzybryHvsxNrra:-KiaU8y_36HHLmPcLRHE:b-z7q5af) ([raw file](https://github.com/kubernetes/kubernetes/blob/9c868e2f987d9d55ed4a19cc81c4cec35dc83ded/federation/pkg/federation-controller/service/dns.go#L250)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

> where the client doing discovery gets a stale 'A' Record. We are leaving the burden to client to figure out its not healthy and move on to another record.

I am not entirely sure I understand. Could you please elaborate?
</blockquote></details>

Say we have federation of 3 clusters, and deploy a federated service. We would be creating a global level DNS records with 3 'A' records. when any client looksup this name, one of these 3 is returned. overtime if one of the cluster goes down(unreachable), the record pertaining to that cluster for the service should be removed. otherwise client may get this stale record and have to realize that its not healthy and use another record.

@madhusudancs, i realized that this case is happening in current implementation (without this pr) too. so please do ignore this comment for now.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

madhusudancs commented:
> Say we have federation of 3 clusters, and deploy a federated service. We would be creating a global level DNS records with 3 'A' records. when any client looksup this name, one of these 3 is returned. overtime if one of the cluster goes down(unreachable), the record pertaining to that cluster for the service should be removed. otherwise client may get this stale record and have to realize that its not healthy and use another record.

Is this specific to CoreDNS implementation? The way this works in most implementations of DNS is by having a single A record with 3 addresses. For reference here is how Google Cloud DNS record looks like:

```JSON
{
 "rrsets": [
  {
   "kind": "dns#resourceRecordSet",
   "name": "federated-ingress-service.e2e-tests-federated-ingress-cqfp0.e2e-federation.svc.pr-bldr.test-f8n.k8s.io.",
   "type": "A",
   "ttl": 180,
   "rrdatas": [
    "107.178.213.149",
    "104.198.218.219",
    "107.178.218.26"
   ]
  }
 ]
}
```

If you want to be consistent with this interface here, what you need to do in the rrsets.Get(name) method is to collect all the records with the same name and type A into a single record as opposed to creating a new record for each node returned here - https://github.com/madhusudancs/kubernetes/blob/222b24ffa13eb26206eb77b43055c7a89cb04669/federation/pkg/dnsprovider/providers/coredns/rrsets.go#L78

Does that make sense?

madhusudancs commented:
@csbell PTAL.

csbell commented:
/lgtm

madhusudancs committed:
Remove all the existing records before creating new ones to avoid DNS misconfiguration.

When we fetch the dns records by name, we get a list of records that match
the given name. As an optimization we look up to see if the new record we
want to create is already in the returned list to avoid performing any updates.

However, when the new record we want to create isn't in the returned list, it
is hard to say if the returned list contains the list of records that we want
to retain. For example, we might get a list of A records and we want to create
a CNAME record. Creating a new CNAME record without removing the A records is
a DNS misconfiguration. So to play safe we just remove all the existing records
in the list and create the new desired record.

**Note**: This is the opposite of what I said here - https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44626#-Ki9xQOzybryHvsxNrra.

madhusudancs commented:
@shashidharatd @csbell PTAL the last commit which I pushed now. I have done the opposite of what I argued here - https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44626#-Ki9xQOzybryHvsxNrra, in other words I have done what @shashidharatd proposed initially. Sorry about that, but please see the commit description for why I changed my mind.

---

Review status: 3 of 12 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-:-KjnzUgde2yoZid9NwUo:b-i8myhk)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
Reviewed 1 of 1 files at r4.
Review status: 4 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 298 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-KjpwB92MmA4DXROfyMa:-KjpwB92MmA4DXROfyMb:bylyyfp) ([raw file](https://github.com/kubernetes/kubernetes/blob/24173e92c240f8feb419e34f26eb95388b6ec692/federation/pkg/federation-controller/service/dns.go#L298)):*
> ```Go
> 				}
> 				changeSet = changeSet.Add(newRrset)
> 				if err = changeSet.Add(newRrset).Apply(); err != nil {
> ```

@madhusudancs, you are adding newRrset twice to the changeSet. please correct.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Review status: 3 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 298 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-KjpwB92MmA4DXROfyMa:-KjqUE2HBwjMJ_ekgdap:bjjq4rc) ([raw file](https://github.com/kubernetes/kubernetes/blob/24173e92c240f8feb419e34f26eb95388b6ec692/federation/pkg/federation-controller/service/dns.go#L298)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

@madhusudancs, you are adding newRrset twice to the changeSet. please correct.
</blockquote></details>

@shashidharatd thanks for catching that. Removed. PTAL.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
/lgtm
@madhusudancs, Thanks for addressing the concerns.

---

Reviewed 2 of 2 files at r3, 1 of 1 files at r5.
Review status: 6 of 12 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44626#-:-Kjqzxt457U2rMnEMFMB:b2182zi)*
<!-- Sent from Reviewable.io -->
